external-crates/move/move-compiler/src/editions/mod.rs

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module controls feature gating and breaking changes in new editions of the source language
+
+use std::{fmt::Display, str::FromStr};
+
+use crate::{diag, shared::CompilationEnv};
+use move_ir_types::location::*;
+use serde::{Deserialize, Serialize};
+
+//**************************************************************************************************
+// types
+//**************************************************************************************************
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, PartialOrd, Ord)]
+pub enum Edition {
+    Legacy,
+    E2024(EditionRelease),
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, PartialOrd, Ord)]
+pub enum EditionRelease {
+    Final,
+    Beta,
+    Alpha,
+}
```

Comment:
This seems a bit rigid to me? What if we want to have multiple alphas/betas, etc? Thoughts on a more relaxed representation, e.g. `struct Edition(u16, Option<String>)` [1]?  I do think it's still worth having a binary be aware of all the editions it supports, but I'm not sure it's helpful to be so strict with the parsing of it.

[1]: This requires `Legacy` to be represented by a `None` (so in the manifest we get an `Option<Edition>`.

Reply:
I like some parts of this, but I worry about the lack of enumeration on the `u16` and `Option<String>`. 
How would you set the feature gates with the stringly typed releases?

Maybe more broadly, I'm not exactly sure what case we are solving here by opening up the representation?

Reply:
> Maybe more broadly, I'm not exactly sure what case we are solving here by opening up the representation?

When it comes to the edition release, it's a case of wanting the flexibility to allow for multiple different names, e.g. we may want to release an `alpha-1`, `alpha-2` etc, and when we do that, I don't think it makes sense to add that to a list of all possible releases.  The equivalent part in [semver](https://semver.org/#spec-item-9) is a very flexible label.

For the year, I'm just trying to avoid the boilerplate -- enumeration is easily solved by having a list of versions that the binary supports, I'm not sure that apart from that it's valuable to enumerate all the editions alone?